test/shared/src/main/scala-2.x/zio/test/mock/Macros.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2019-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test.mock
+
+import scala.reflect.macros.blackbox.Context
+
+import zio.Has
+
+private[test] object Macros {
+
+  def summonMockable_impl[R <: Has[_]: c.WeakTypeTag](c: Context): c.Expr[Mockable[R]] = {
```

Comment:
This is the first instance of using macros in Core... if we're going to do this for 2.x we should make sure there's a corresponding path forward for Dotty.

Reply:
We actually use a them already to implement the `assertCompiles` functionality, though that is fairly specialized and Dotty provides specific support for that.

Reply:
Well, the `typeChecks` assertion was already using macros in `zio.test` (and this one is also in `zio.test`) and there was really no other way (implicit induction won't work here unfortunately), thats why I decided to introduce it.

For now the effect on Dotty is -> you'd have to create the composed `Mockable` instanced on your own. The `MockableDerivation` is lowest in the implicit search, so if you do provide it (by any other means) the disfunctional/missing macro in Dotty won't affect you.

If you do nothing, currently Dotty auto-derrivation will throw compile-time error.

Reply:
I have not yet went deeply into Dotty, but AFAIK some features of meta-programming for Dotty are not yet available.